datalad_next/runners/data_processors.py

```diff
@@ -0,0 +1,316 @@
+    return parts_with_ends, []
+
+
+def _pattern_border_processor(pattern: StrOrBytes,
+                              data_chunks: StrOrBytesList,
+                              final: bool = False,
+                              ) -> tuple[StrOrBytesList, StrOrBytesList]:
+
+    def ends_with_pattern_prefix(data: StrOrBytes, pattern: StrOrBytes) -> bool:
+        """ Check whether the chunk ends with a prefix of the pattern """
+        for index in range(len(pattern) - 1, 0, -1):
+            if  data[-index:] == pattern[:index]:
+                return True
+        return False
+
+    # Copy the list, because we might modify it and the caller might not expect that.
+    data_chunks = data_chunks[:]
+
+    # Join data chunks until they are sufficiently long to contain the pattern,
+    # i.e. have a least the size: `len(pattern)`. Continue joining, if the chunk
+    # ends with a prefix of the pattern.
+    current_index = 0
+    while current_index < len(data_chunks) - 1:
+        current_chunk = data_chunks[current_index]
+        while (len(data_chunks[current_index:]) > 1
+               and (len(current_chunk) < len(pattern)
+                    or ends_with_pattern_prefix(current_chunk, pattern))):
+            data_chunks[current_index] += data_chunks[current_index + 1]
+            del data_chunks[current_index + 1]
+            current_chunk = data_chunks[current_index]
+        current_index += 1
+
+    # At this point we have joined whatever we can join. We still have to check
+    # whether the last chunk ends with a pattern-prefix.
+    if not final:
+        if ends_with_pattern_prefix(data_chunks[-1], pattern):
+            return data_chunks[:-1], data_chunks[-1:]
+    return data_chunks, []
+
+
+class SplitLinesProcessor:
+    def __init__(self,
+                 separator: str | None = None,
+                 keep_ends: bool = False
+                 ):
+        self.line_splitter = LineSplitter(separator, keep_ends)
+
+    def __call__(self,
+                 data_chunks: StrOrBytesList,
+                 final: bool = False,
+                 ) -> tuple[StrOrBytesList, StrOrBytesList]:
+        result = [
+            line
+            for data in data_chunks
+            for line in self.line_splitter.process(data)
+        ]
+        if final:
+            remaining = self.line_splitter.finish_processing()
+            if remaining:
+                result.append(remaining)
+        return result, []
+
+
+class DecodeProcessor:
+    def __init__(self, encoding: str = 'utf-8'):
+        self.encoding = encoding
+
+    def __call__(self, data_chunks: BytesList,
+                 final: bool = False
+                 ) -> tuple[StrList, BytesList]:
+        try:
+            text = (b''.join(data_chunks)).decode(self.encoding)
+        except UnicodeDecodeError:
+            if final:
+                raise
+            return [], data_chunks
+        return [text], []
+
+
+class ProcessorPipeline:
```

Comment:
Wondering what the envisioned *public* API would be here. It seems `process()` and `finalize()` would/should not be called outside of `process_from()`. If that is true, could `process_from()` become `__call__()`?

Reply:
`process` and `finalize` could be called by protocol implementations in their `pipe_data_received` and `connection_lost`-callbacks. This could be especially useful for non-generator protocols. I think it would help us to reduce code duplication in those protocols and robustify the data handling.

But we could still make `process_from` a `__call__`.